applications/OptimizationApplication/python_scripts/utilities/opt_convergence.py

```diff
@@ -81,3 +85,97 @@ def GetInfo(self) -> dict:
                 'tolerance': self.__tolerance,
                 'status': str("converged" if self.conv else "not converged")}
         return info
+
+class AverageAbsoluteImprovement:
+    @classmethod
+    def GetDefaultParameters(cls):
+        return Kratos.Parameters("""{
+            "type"              : "aver_abs_delta",
+            "max_iter"          : 0,
+            "tracked_iter"      : 5,
+            "tolerance"         : 1e-6
+        }""")
+
+    def __init__(self, parameters: Kratos.Parameters, optimization_problem: OptimizationProblem):
+        parameters.ValidateAndAssignDefaults(self.GetDefaultParameters())
+        self.__max_iter = parameters["max_iter"].GetInt()
+        self.__tracked_iter = parameters["tracked_iter"].GetInt()
+        self.__optimization_problem = optimization_problem
+        self.__tolerance = parameters["tolerance"].GetDouble()
+        self.__abs_value_change = []
+        self.value = 9999
+
+    @time_decorator()
+    def IsConverged(self) -> bool:
+        iter = self.__optimization_problem.GetStep()
+        self.conv = iter >= self.__max_iter
+
+        step = self.__optimization_problem.GetStep()
+        print(step)
+        algorithm_buffered_data = ComponentDataView("algorithm", self.__optimization_problem).GetBufferedData()
+        if step == 0:
+            self.__init_value = algorithm_buffered_data.GetValue("std_obj_value", 0)
+            self.conv = False
+        if step > 0:
+            if not algorithm_buffered_data.HasValue("std_obj_value"):
+                raise RuntimeError(f"Algorithm data does not contain computed \"std_obj_value\".\nData:\n{algorithm_buffered_data}" )
+            value = algorithm_buffered_data.GetValue("std_obj_value", 0)
+            prev_value = algorithm_buffered_data.GetValue("std_obj_value", 1)
+            self.__abs_value_change.append((value - prev_value)/self.__init_value)
+
+        if step >= self.__tracked_iter - 1:
+            self.value = 0.0
+            for i in range(self.__tracked_iter - 1):
+                self.value += self.__abs_value_change[-1 - i]
+            self.value = abs(self.value) / (self.__tracked_iter - 1) 
+            self.conv = self.value <= self.__tolerance
+
+        DictLogger("Convergence info",self.GetInfo())
+
+        return self.conv
+
+    def GetInfo(self) -> dict:
+        info = {'type': 'aver_abs_delta',
+                'aver_abs_delta [%]': self.value * 100,
+                'tolerance [%]': self.__tolerance * 100,
+                'tracked_iter': self.__tracked_iter,
+                'status': str("converged" if self.conv else "not converged")}
+        return info
```

Comment:
this seems like an ideal application for a `dataclass`:

```py
@dataclass
class Info:
    type_info: str
    aver_abs_delta: float
    # ...
```
this makes it type safe

```

Reply:
We need to have properly defined interface class for `ConvergenceCriteria`, in this case the data class would not be possible ryt?

Reply:
a dataclass is just a class in the end ...
you can add methods and everything to it

Having said this, if you want to leave the `status` var, it should probably also be an enum => `CONVERGED` & `NOT_CONVERGED`